refactor!: NetworkPrefabHandler to replace Custom Spawn/Destroy Handlers

NoelStephensUnity opened the pull request:
**This PR is directly tied to [MTT-622](https://jira.unity3d.com/browse/MTT-622)** 
Due to the changes in the GlobalObjectHashId, a refactor the custom spawn and destroy handlers for NetworkSpawnManager is required.
In order to maintain the same functionality/features while taking future architectural changes into consideration, a new class called NetworkPrefabHandler will be created.
This PR also restores the ConnectionModeButtons.prefab being deleted in [PR-694](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/pull/694/files)

NoelStephensUnity committed:
refactor!: Prefab Spawn Handling

This PR isolates all prefab spawn handling into its own class.
Registration is handled via the NetworkManager.PrefabHandler.
Using one interface as opposed to two delegate handlers.

NoelStephensUnity committed:
Adding Unit Test for NetworkPrefabHandler

This is 99% NetworkPrefabHandler unit test and 1% fixing a misspelled word in RpcQueueTests.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/networkprefabhandler

NoelStephensUnity commented on com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs:
```diff
@@ -333,7 +333,7 @@ public static void NetworkHide(List<NetworkObject> networkObjects, ulong clientI
         private void OnDestroy()
         {
-            if (NetworkManager.Singleton != null && NetworkManager.Singleton.SpawnManager.SpawnedObjects.ContainsKey(NetworkObjectId))
+            if (NetworkManager.Singleton != null && NetworkManager.Singleton.SpawnManager != null && NetworkManager.Singleton.SpawnManager.SpawnedObjects.ContainsKey(NetworkObjectId))
```
This fixes an issue where it is possible to be destroying after the NetworkSpawnManager no longer exists.

MFatihMAR replied:
@TwoTenPvP might also be interested while he's working on singleton removal tasks these days.

NoelStephensUnity committed:
update commentd

Updated the comments for the NetworkPrefabHandler

MFatihMAR commented on com.unity.multiplayer.mlapi/Tests/Runtime/NetworkPrefabHandlerTests.cs:
```diff
@@ -0,0 +1,162 @@
+            Assert.False(networkPrefaInstanceHandler.StillHasInstances());
+        }
+
+        public void Dispose()
+        {
+            //Stop, shutdown, and destroy
+            NetworkManagerHelper.ShutdownNetworkManager();
+        }
+
+        public NetworkPrefabHandlerTests()
+        {
+            //Create, instantiate, and host
+            NetworkManagerHelper.StartNetworkManager();
+        }
```
you could also use `[SetUp]` and `[TearDown]` attributed methods instead of `class ctor` and `Dispose()`
- https://docs.unity3d.com/Packages/com.unity.test-framework@1.1/manual/reference-actions-outside-tests.html
- https://docs.nunit.org/articles/nunit/technical-notes/usage/SetUp-and-TearDown.html

NoelStephensUnity replied:
SetUpAttribute **is now used exclusively for per-test setup**.
TearDownAttribute **is now used exclusively for per-test teardown**.

The NetworkManagerHelper method for unit tests **is meant to stay active during the duration of all tests selected** to be run.  While this is one test, there are many other tests that could be run that might need to use the NetworkManager instance created for the temporary test scene and I also plan on adding more tests to NetworkPrefabHandlerTests.

The reason for the change in the setup and tear down is to deal with scenarios where you allocate something but have an assert that breaks the current stack scope (_i.e. if you tried to allocate something that needs to be disposed/destroyed during a test and expected to do these actions at the end of your test but triggered an Assert...you would never dispose/destroy said allocated objects...thus a per test setup and tear down pattern_)

NoelStephensUnity replied:
Replaced the constructor and dispose methods with Setup and TearDown attributes.
Made changes to NetworkManagerHelper to handle the remaining issues with instantiating and destroying NetworkManager multiple times.

NoelStephensUnity committed:
style:  GlobalObjectHashId to GlobalObjectIdHash

fixing typo.

NoelStephensUnity committed:
refactor: Unit test Initialization and Asset Recovery

Removed singleton references from NetworkManagerHelper in order to prepare for this as well as fix an issue with destroying and instantiation NetworkManager instances during unit tests.

Added the ConnectionModeButtons prefab back for the manual RPC Testing tool that was deleted in PR 694
https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/pull/694/files
